### PR TITLE
kubectl: enhance podtemplate describer

### PR DESCRIPTION
**Pull request by 0xmichalis**
@kubernetes/kubectl 

@smarterclayton for the dc decsriber

### CODE REVIEW

**Comment by kargakis** [comment]
Instead of exporting the volume describer I am looking into packing everything in the podtemplate describer...

**Comment by kargakis** [comment]
cc: @janetkuo since the only place I can see the podtemplate describer used is `kubectl rollout history`. We want to use it in our DeploymentConfigs, and we should also use it for Deployments. PTAL

**Comment by bgrant0607** [comment]
ref #20941
cc @AdoHe

**Comment by smarterclayton** [comment]
Changes look good to me, but want a second set of eyes.

**Comment by AdoHe** [comment]
I will take a deep look.

**Comment by kargakis** [comment]
@AdoHe ping

**Comment by janetkuo** [comment]
Current history output looks like this:
```console
$ kubectl rollout history deployment --revision=1
deployments "nginx-deployment" revision 1
  Labels:	app=nginx,pod-template-hash=1564180365
  Annotations:	kubernetes.io/change-cause=kubectl create -f docs/user-guide/new-nginx-deployment.yaml --record
  Containers:
  nginx:
    Image:	nginx:1.9.1
    Port:	80/TCP
    QoS Tier:
      cpu:	BestEffort
      memory:	BestEffort
    Environment Variables:
  No volumes.
```

But there should be an indention ("  ") for each individual container, like how we describe pods:
```console
Name:		nginx-deployment-1564180365-tirqi
Namespace:	default
...
Containers:
  nginx:
    Container ID:	docker://9c526755a2433159af940caeeeaa9b7b3b6123501004251de9766fcd1324e170
    Image:		nginx:1.9.1
...
```

An easy fix is to change the output of history to:
```console
$ kubectl rollout history deployment --revision=1
Name:			nginx-deployment
Revision:		1
Labels:			app=nginx,pod-template-hash=1564180365
Annotations:	kubernetes.io/change-cause=kubectl create -f docs/user-guide/new-nginx-deployment.yaml --record
Containers:
  nginx:
    Image:	nginx:1.9.1
    Port:	80/TCP
    QoS Tier:
      cpu:	BestEffort
      memory:	BestEffort
    Environment Variables:
No volumes.
```

**Comment by janetkuo** [comment]
Also can you change `describeContainers` to hide `Environment Variables:` or show `<none>` when there's no env variable?

File: pkg/kubectl/describe.go [hunk 1]
```diff
@@ -992,18 +992,21 @@ func describeReplicationController(controller *api.ReplicationController, events
...
-		fmt.Fprintf(out, "Image(s):\t%s\n", makeImageList(&template.Spec))
-		describeVolumes(template.Spec.Volumes, out)
-		return nil
-	})
+func DescribePodTemplate(template *api.PodTemplateSpec, out io.Writer) {
+	if template == nil {
+		fmt.Fprintf(out, "  <unset>")
+		return
+	}
+	fmt.Fprintf(out, "  Labels:\t%s\n", labels.FormatLabels(template.Labels))
```
Comment (janetkuo) [inline pkg/kubectl/describe.go]: `if len(template.Labels) > 0` here too?

**Comment by janetkuo** [comment]
In pkg/kubectl/history.go#History, can you change 
```go
historyInfo.RevisionToTemplate[v].Annotations[ChangeCauseAnnotation] = changeCause
```
to 
```go
if len(changeCause) > 0 {
	historyInfo.RevisionToTemplate[v].Annotations[ChangeCauseAnnotation] = changeCause
}
```

File: pkg/kubectl/describe.go [hunk 2]
```diff
@@ -766,8 +766,8 @@ func (d *PersistentVolumeClaimDescriber) Describe(namespace, name string) (strin
 	})
 }
 
-// DescribeContainers is exported for consumers in other API groups that have container templates
-func DescribeContainers(containers []api.Container, containerStatuses []api.ContainerStatus, resolverFn EnvVarResolverFunc, out io.Writer) {
+// describeContainers is exported for consumers in other API groups that have container templates
+func describeContainers(containers []api.Container, containerStatuses []api.ContainerStatus, resolverFn EnvVarResolverFunc, out io.Writer) {
```
Comment (AdoHe) [inline pkg/kubectl/describe.go]: @smarterclayton IIRC you add this function at begin, is that ok to make this function non pubic?

File: pkg/kubectl/describe.go (hunk 1 above)
Comment (AdoHe) [inline pkg/kubectl/describe.go]: I think `labels.FormatLabels` will handle `len(template.Labels) == 0`

File: pkg/kubectl/describe.go [hunk 3]
```diff
@@ -992,18 +992,21 @@ func describeReplicationController(controller *api.ReplicationController, events
...
-		return nil
-	})
+func DescribePodTemplate(template *api.PodTemplateSpec, out io.Writer) {
+	if template == nil {
+		fmt.Fprintf(out, "  <unset>")
+		return
+	}
+	fmt.Fprintf(out, "  Labels:\t%s\n", labels.FormatLabels(template.Labels))
+	if len(template.Annotations) > 0 {
+		fmt.Fprintf(out, "  Annotations:\t%s\n", labels.FormatLabels(template.Annotations))
```
Comment (AdoHe) [inline pkg/kubectl/describe.go]: also for annotations display, we may need special handling, you can keep this here, I will update this later.

**Comment by kargakis** [comment]
> And then you don't need space string in describeVolumes anymore.

The problem here is that the podtemplate describer needs to indent all of its children. It shows containers, volumes, template labels, annotations, etc. If we are going to have them as top level info then labels and annotations will clash with the object meta labels/annotations. I think of `kubectl rollout history` as a special case so maybe it deserves a separate describer.

File: pkg/kubectl/describe.go (hunk 1 above)
Comment (kargakis) [inline pkg/kubectl/describe.go]: it's common for podtemplates to have labels, that's why I opted on showing the label field all the time. I can do the same for annotations but podtemplate annotations are not so common.

File: pkg/kubectl/describe.go (hunk 2 above)
Comment (kargakis) [inline pkg/kubectl/describe.go]: @AdoHe yes now we are going to use it via the podtemplate describer so I think it's fine

**Comment by kargakis** [comment]
I have added an extra space for {container,volume} names in case we are trying to describe a podtemplate. I believe all objects containing podtemplates should start using the podtemplate describer. Thus, everything will be consistent and we will stop doing the indent hacks I have in this PR. Spawned https://github.com/kubernetes/kubernetes/issues/23429.

**Comment by kargakis** [comment]
> Spawned #23429.

https://github.com/kubernetes/kubernetes/issues/20941 supersedes #23429

**Comment by janetkuo** [comment]
LGTM. @AdoHe may have more comments.

**Comment by janetkuo** [comment]
unit test fail:
```
# k8s.io/kubernetes/pkg/kubectl
./describe_test.go:271: not enough arguments in call to describeContainers
FAIL	k8s.io/kubernetes/pkg/kubectl [build failed]
```

**Comment by kargakis** [comment]
squashed down to one commit, unit test failure fixed

@AdoHe PTAL

**Comment by AdoHe** [comment]
LGTM.

**Comment by kargakis** [comment]
@janetkuo can you please tag this appropriately? Thanks!

**Pull request merged**